### PR TITLE
Add & Fix examples, docs

### PR DESCRIPTION
**Pull request by luxfg**
I fixed some the examples and added two simple vertical stencils, a forward and a backward, and a 4 PE 2-phase reduce.

### CODE REVIEW

**Comment by luxfg** [comment]
Note that tests are failing because of issues in the parsing for hops and for-loops. Could you investigate?

File: irspec/docs/spatial/examples.md [hunk 1]
```diff
@@ -183,18 +168,10 @@ kernel laplacian<I,J,K> (stream<f32>[I+2, J+2] readonly in_field,
...
-  
-    dataflow i16 i, i16 j in [0:I+2, 0:J+2] {
-      stream<f32> in_field_l = in_field[i, j];
-    }
-  
     compute i16 i, i16 j in [0:I+2, 0:J+2] {
-      foreach k, x in [0:K], receive(in_field_l) {
-        local_input[k] = x;
-      }
+      receive(lap_field[i, j], local_input);
```
Comment (tbennun) [inline irspec/docs/spatial/examples.md]: inverted arguments?

File: irspec/docs/spatial/examples.md [hunk 2]
```diff
@@ -150,8 +135,8 @@ kernel vadv<I,J,K>(stream<f32>[I, J] utens_stage,
         }
   
         // Copy the data to the output
-        send(datacol_l, datacol_c);
-        send(utens_stage_l, utens_stage_c);
+        send(datacol_l, datacol[i, j]);
```
Comment (tbennun) [inline irspec/docs/spatial/examples.md]: correct syntax should always be either `await send` or `completion c = send`, applies to all the other examples.

File: irspec/docs/spatial/examples.md [hunk 3]
```diff
@@ -329,42 +295,47 @@ kernel <I, J, K>hdiff(stream<f32>[I, J] readonly in_stream,
...
-        }
-        
-        await c;
-
-        comp c2 = send(local_input, eastwards);
-        
-        // result not needed, decrement directly.
-        await foreach i32 k, f32 x in [0:K], receive(eastwards) {
-            lap_field[k] = lap_field[k] - x;
+        dataflow i16 i, i16 j in [1:I, 1:J] {
```
Comment (tbennun) [inline irspec/docs/spatial/examples.md]: computation in dataflow

File: irspec/docs/spatial/examples.md [hunk 4]
```diff
@@ -484,22 +455,22 @@ kernel conv<J>(stream<f32>[J] readonly input,
             y = x * kernel[1];
 
             // Send the data to the right
-            comp_east = send(x, eastwards);
+            comp_west = send(x, westwards);
```
Comment (tbennun) [inline irspec/docs/spatial/examples.md]:
```suggestion
            completion comp_west = send(x, westwards);
```

File: irspec/docs/spatial/examples.md [hunk 5]
```diff
@@ -484,22 +455,22 @@ kernel conv<J>(stream<f32>[J] readonly input,
...
+            await foreach i16 k, f32 x2 in [0:1], receive(westwards) {
               y = y + x2 * kernel[0];
             }
 
-            await comp_east;
+            await comp_west;
 
             // Send the data to the left
-            comp_west = send(x, westwards);
+            comp_east = send(x, eastwards);
```
Comment (tbennun) [inline irspec/docs/spatial/examples.md]: same as above

File: irspec/docs/spatial/spatial.md [hunk 6]
```diff
@@ -580,6 +595,24 @@ After the completion triggers, the stream may be used for other sends or receive
     *Failure to uniquely match nested `send`s and `receives` results
     in incorrect stream edges and raises a compilation error.*
 
+### Receive Statement
```
Comment (tbennun) [inline irspec/docs/spatial/spatial.md:598]: thx for adding these two elements!!

**Review by tbennun — CHANGES_REQUESTED** [review]
Looks good, see comments

**Comment by luxfg** [comment]
Ok, I fixed the examples and the parser & ir.

**Review by tbennun — APPROVED** [review]

**Pull request merged**